hal/hal_remote/include/hal_remote/hal_binary_decoder.h

```diff
@@ -0,0 +1,388 @@
+        return sizeof(message.write.dst) + sizeof(message.write.size);
+      case hal_binary_encoder::COMMAND::MEM_READ:
+        return sizeof(message.read.src) + sizeof(message.read.size);
+      case hal_binary_encoder::COMMAND::MEM_FILL:
+        return sizeof(message.fill.dst) + sizeof(message.fill.pattern_size) +
+               sizeof(message.fill.size);
+      case hal_binary_encoder::COMMAND::MEM_COPY:
+        return sizeof(message.copy.dst) + sizeof(message.read.src) +
+               sizeof(message.read.size);
+      case hal_binary_encoder::COMMAND::PROGRAM_FREE:
+        return sizeof(message.prog_free.program);
+      case hal_binary_encoder::COMMAND::PROGRAM_LOAD:
+        return sizeof(message.prog_load.size);
+      case hal_binary_encoder::COMMAND::FIND_KERNEL:
+        return sizeof(message.prog_find_kernel.kernel_name_size) +
+               sizeof(message.prog_find_kernel.program);
+      case hal_binary_encoder::COMMAND::MEM_ALLOC_REPLY:
+        return sizeof(message.alloc_reply);
+      case hal_binary_encoder::COMMAND::PROGRAM_LOAD_REPLY:
+        return sizeof(message.prog_load_reply);
+      case hal_binary_encoder::COMMAND::KERNEL_EXEC:
+        return sizeof(message.kernel_exec.program) +
+               // 8
+               sizeof(message.kernel_exec.kernel) +
+               // 16
+               sizeof(message.kernel_exec.num_args) +
+               // 20
+               sizeof(message.kernel_exec.nd_range.offset[0]) * 9 +
+               // 56
+
+               sizeof(message.kernel_exec.work_dim) +
+               sizeof(message.kernel_exec.args_data_size);
+        // 64
+      case hal_binary_encoder::COMMAND::DEVICE_CREATE:
+      case hal_binary_encoder::COMMAND::DEVICE_DELETE:
+        return 0;
+      case hal_binary_encoder::COMMAND::FIND_KERNEL_REPLY:
+        return sizeof(message.find_kernel_reply);
+      case hal_binary_encoder::COMMAND::MEM_READ_REPLY:
+      case hal_binary_encoder::COMMAND::MEM_WRITE_REPLY:
+      case hal_binary_encoder::COMMAND::MEM_FILL_REPLY:
+      case hal_binary_encoder::COMMAND::MEM_FREE_REPLY:
+      case hal_binary_encoder::COMMAND::MEM_COPY_REPLY:
+      case hal_binary_encoder::COMMAND::PROGRAM_FREE_REPLY:
+      case hal_binary_encoder::COMMAND::KERNEL_EXEC_REPLY:
+      case hal_binary_encoder::COMMAND::DEVICE_CREATE_REPLY:
+      case hal_binary_encoder::COMMAND::DEVICE_DELETE_REPLY:
+        return sizeof(uint32_t);  // bool as 32 bit
+        break;
+      default:
+        printf("Unknown command : %d\n", (int)command);
+        assert(0 && "Command cannot be decoded");
+    }
+    return data_required_unknown;
+  }
+
+  /// @brief  decode a command given data of a particular sise
+  /// @param command
+  /// @param data
+  /// @param size length of data in bytes
+  /// @return true if succeeded otherwise false.
+  bool decode(hal_binary_encoder::COMMAND command, const uint8_t *data,
+              uint32_t size) {
```

Comment:
Can we use `cargo::result` here to make it clearer from the API what the meaning of the result is?

Reply:
Also, I cannot really follow the logic of when `uint8_t *` is used vs. `void *`, I would have expected `void *` here but it's not a problem if there is a reason for it.

Reply:
I don't want to drag in cargo, we've avoided it in hal so far in keeping it mostly standalone (and working with clik).
I think you are right about void * though, although we do tend to need to then cast it elsewhere to index into it.